src/Cryptography/Sodium/Bindings/SecureMemory.hs

```diff
@@ -0,0 +1,91 @@
+{-# LANGUAGE CApiFFI #-}
+{-# LANGUAGE Safe #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+--
+-- Module: Cryptography.Sodium.Bindings.SecureMemory
+-- Description: Direct bindings to the libsodium secure memory functions
+-- Copyright: (C) Hécate Moonlight 2022
+-- License: BSD-3-Clause
+-- Maintainer: The Haskell Cryptography Group
+-- Stability: Stable
+-- Portability: GHC only
+module Cryptography.Sodium.Bindings.SecureMemory
+  ( -- * Introduction
+    -- $introduction
+
+    -- * Zeroing memory
+    memZero,
+
+    -- * Locking memory
+    lock,
+    unlock,
+  )
+where
+
+import Foreign (Ptr)
+import Foreign.C.Types (CInt (CInt), CSize (CSize))
+
+-- $introduction
+-- This module provides bindings to the secure memory functions provided by Libsodium.
+-- It is intended to be qualified on import:
+--
+-- > import qualified Cryptography.Sodium.Bindings.SecureMemory as SecureMemory
+--
+-- It is recommended to disable swap partitions on machines processing sensitive
+-- data or, as a second choice, use encrypted swap partitions.
+--
+-- For similar reasons, on Unix systems, one should also disable core dumps when
+-- running crypto code outside a development environment.
+-- This can be achieved using a shell built-in such as @ulimit@ or programmatically
+-- using @setrlimit(RLIMIT_CORE, &(struct rlimit) {0, 0})@.
+-- On operating systems where this feature is implemented, kernel crash dumps
+-- should also be disabled.
+--
+-- The 'lock' function wraps @mlock()@ and @VirtualLock()@.
+--
+-- Note: Many systems place limits on the amount of memory that may be locked
+-- by a process. Care should be taken to raise those limits (e.g. Unix ulimits)
+-- where necessary.
+
+-- | Overwrite the memory region starting at the pointer
+-- address with zeros.
+--
+-- @since 0.0.1.0
+foreign import capi "sodium.h sodium_memzero"
+  memZero ::
+    -- | Start pointer
+    Ptr x ->
```

Comment:
This should be a pointer to `Word8` in my opinion, as you're wiping a sequence of _bytes_ specifically.